---
title: "Session 1: containerising a pipeline"
teaching: 10
exercises: 50
questions:
objectives:
---


### Get ready for the workshops

By now, you'll be connected to the Nimbus VM you got assigned to.  

First thing, we need to download the workshop materials from Github:

```
$ cd /data
$ git clone https://github.com/PawseySC/containers-bioinformatics-workshop.git
$ cd containers-bioinformatics-workshop
$ export WORK=$(pwd)
```
{: .bash}

To save you some potentially long download times, we have cached for you most of the container images required by this workshop; we've done it with the script `$WORK/exercises/pull_images.sh`.


### Goals

In this first session, we're going to port a small RNA sequencing pipelines to containers.  Here is a graphical representation:

<!-- ![Pipeline DAG]({{ page.root }}/fig/pipeline_dag.png) -->
<img src="{{ page.root }}/fig/pipeline_dag.png" alt="Pipeline DAG" width="485" height="525"/>

The pipeline uses three tools, `salmon`, `fastqc` and `multiqc`.

By containerising it, we're going to apply some skills we learnt about in the webinar, namely:

* search for container images on web registries;
* download images with `singularity pull <IMAGE>`
* execute commands in containers through `singularity exec <IMAGE> <CMD> <ARGS>`
* bind mount additional host directories using either
  - execution flag `-B`
  - environment variable `SINGULARITY_BINDPATH`

If you need to brush up on these concepts, have a look at these webinar episodes:

* [Basic of Singularity](https://pawseysc.github.io/singularity-containers/12-singularity-intro/index.html)

* [Share files with the host: BLAST, a bioinformatics demo](https://pawseysc.github.io/singularity-containers/13-bio-example-host/index.html)

Towards the end, we're going to build on these concepts, and see how we can eventually simplify the user experience with containerised packages.


### 