---
title: "Session 1: containerising a pipeline"
teaching: 10
exercises: 50
questions:
objectives:
---


### Get ready for the workshops

By now, you'll be connected to the Nimbus VM you got assigned to.  

First thing, we need to download the workshop materials from Github:

```
$ cd /data
$ git clone https://github.com/PawseySC/containers-bioinformatics-workshop.git
$ cd containers-bioinformatics-workshop
$ export WORK=$(pwd)
```
{: .bash}

To save you some potentially long download times, we have cached for you most of the container images required by this workshop; we've done it with the script `$WORK/exercises/pull_images.sh`.


### Goals

In this first session, we're going to port a small RNA sequencing pipelines to containers.  Here is a graphical representation:

<!-- ![Pipeline DAG]({{ page.root }}/fig/pipeline_dag.png) -->
<img src="{{ page.root }}/fig/pipeline_dag.png" alt="Pipeline DAG" width="485" height="525"/>

The pipeline uses three tools, `salmon`, `fastqc` and `multiqc`.

By containerising it, we're going to apply some skills we learnt about in the webinar, namely:

* search for container images on web registries;
* download images with `singularity pull <IMAGE>`
* execute commands in containers through `singularity exec <IMAGE> <CMD> <ARGS>`
* bind mount additional host directories using either
  - execution flag `-B`
  - environment variable `SINGULARITY_BINDPATH`

If you need to brush up on these concepts, have a look at these webinar episodes:

* [Basic of Singularity](https://pawseysc.github.io/singularity-containers/12-singularity-intro/index.html)

* [Share files with the host: BLAST, a bioinformatics demo](https://pawseysc.github.io/singularity-containers/13-bio-example-host/index.html)

Towards the end, we're going to build on these concepts, and see how we can eventually simplify the user experience with containerised packages.


### Get the container images ready

Let's cd in the directory for this exercise:

```
$ cd exercises/pipeline/data
```
{: .bash}

and have a look at the pipeline:

```
$ cat original_pipe.sh
```
{: .bash}

```
#!/bin/bash

echo "Pipeline started..."

# step 1
cd ../reference
# ONLY CHANGE THE NEXT LINE - EXECUTION LINE
salmon index -t ggal_1_48850000_49020000.Ggal71.500bpflank.fa -i out_index &>log_index
cd ../data
echo " indexing completed"

# step 2
# ONLY CHANGE THE NEXT LINE - EXECUTION LINE
salmon quant --libType=U -i ../reference/out_index -1 ggal_gut_1.fq -2 ggal_gut_2.fq -o ggal_gut &>log_quant
echo " quantification completed"

# step 3
mkdir out_fastqc
# ONLY CHANGE THE NEXT LINE - EXECUTION LINE
fastqc -o out_fastqc -f fastq -q ggal_gut_1.fq ggal_gut_2.fq &>log_fq
echo " quality control completed"

# step 4
mkdir out_multiqc
cd out_multiqc
ln -s ../ggal_gut .
ln -s ../out_fastqc .
# ONLY CHANGE THE NEXT LINE - EXECUTION LINE
multiqc -v . &>../log_mq
cd ..
echo " multiple quality control completed"

echo "Pipeline finished!"
```
{: .output}

You can see there's some printing here are there using `echo`, to monitor the progress at runtime.  
Some housekeeping is also being done, such as changing directories and creating output directories.  

The key execution lines are just four though, and make use of three packages:
* salmon
* fastqc
* multiqc

Our first task is to look for container images for this task.  Let's start with Salmon.


> ## Find a container image for Salmon
> 
> Today we're using the web registry `Quay`, at https://quay.io, to search the images we need.  This registry contains all the images provided by the **BioContainers** project, so there are good chances of finding what we need there.  We could have gone directly at the BioContainers home page, https://biocontainers.pro, however its user interface is a bit less friendly right now.
> 
> Now let's try and find the most recent container image for Salmon by BioContainers, using the Quay web page.
> 
> > ## Solution
> > 
> > * Type `salmon` in the *Search* field on the top right of the page; in alternative, use the *Explore* button at the top;
> > * In the results list, look for `biocontainers/salmon` and click on it;
> > * Click on the *Tags* icon on the left, and locate the most recent one at the top of the list; it should be `1.2.1--hf69c8f4_0`;
> > * Click on the *Fetch* icon at the rightmost side of the record, select *Pull by Tag*, and then select and copy to clipboard the string starting with `quay.io/`, *i.e.* `quay.io/biocontainers/salmon:1.2.1--hf69c8f4_0`.
> {: .solution}
{: .challenge}


> ## Pull the container image for Salmon
> 
> To this end let's use the appropriate `singularity` command.
> 
> > ## Solution
> > 
> > ```
> > $ singularity pull quay.io/biocontainers/salmon:1.2.1--hf69c8f4_0
> > ```
> > {: .bash}
> > 
> > At the end an image SIF file for Salmon is downloaded:
> > 
> > ```
> > $ ls salmon*
> > ```
> > {: .bash}
> > 
> > ```
> > salmon_1.2.1--hf69c8f4_0.sif
> > ```
> > {: .output}
> {: .solution}
{: .challenge}


> ## Find and pull images for FastQC and MultiQC
> 
> We're using the same process just adopted for Salmon.
> 
> > ## Solution: FastQC
> > 
> > ```
> > $ singularity pull docker://quay.io/biocontainers/fastqc:0.11.9--0
> > ```
> > {: .bash}
> {: .solution}
> 
> > ## Solution: MultiQC
> > 
> > ```
> > $ singularity pull docker://quay.io/biocontainers/multiqc:1.9--pyh9f0ad1d_0
> > ```
> > {: .bash}
> {: .solution}
{: .challenge}


> ## SIF images in your current directory
> 
> We're in `$WORK/exercises/pipeline/data`.  
> 
> Right now we should have these three images in it, together with the inputs and scripts for the pipeline:
> 
> > ## Solution
> > 
> > ```
> > $ ls
> > ```
> > {: .bash}
> > 
> > ```
> > clean_outputs.sh              ggal_gut_1.fq                 multiqc_1.9--pyh9f0ad1d_0.sif salmon_1.2.1--hf69c8f4_0.sif
> > fastqc_0.11.9--0.sif          ggal_gut_2.fq                 original_pipe.sh
> > ```
> > {: .output}
> {: .solution}
{: .challenge}


### Containerise the pipeline

Now it's your time to work a bit with Singularity!

Make a copy of the original pipeline to work with:

```
$ cp original_pipe.sh pipe.1.sh
```
{: .bash}

At first, you will need to modify ONLY FOUR LINES of this file.  Do not change anything else, nor move files around in the directory structure.  

#### Hints
* add `singularity` syntax to execute those four commands using the appropriate containers you downloaded;
* the current directory is mounted by default in the container;
* if you need other directories, however... (use the flag for binding for now, not the environment variable);
* if you want to go in stages, you can temporarily add an `exit` command in the script after the pipeline step you're working on, to avoid errors by the upcoming steps (not yet containerised).

## Step 1

## Solution

```
singularity exec \
  docker://quay.io/biocontainers/salmon:1.2.1--hf69c8f4_0 \
  salmon index -t ggal_1_48850000_49020000.Ggal71.500bpflank.fa -i out_index &>log_index
```
{: .source}

